File: problems/HackerRank/2521_exercises/linkedList/easy/removeNthFromEnd.c
struct ListNode *removeNthFromEnd(struct ListNode *head, int n)
{

	struct ListNode *dummyNode = (struct ListNode *)malloc(sizeof(struct ListNode));
	dummyNode->next = head;

	// counter the number of nodes
	int count = 0;
	struct ListNode *counter = head;

	while (counter != NULL)
	{
		count++;
		counter = counter->next;
	}

	int target = count - n;
	struct ListNode *current = dummyNode;

	while (target != 0)
	{
		current = current->next;
		target--;
	}

	current->next = current->next->next;

	return dummyNode->next;
}

struct ListNode *removeNthFromEnd(struct ListNode *head, int n)
{

	struct ListNode *dummyNode = (struct ListNode *)malloc(sizeof(struct ListNode));
	dummyNode->next = head;
	struct ListNode *left = dummyNode;
	struct ListNode *right = head;

	while (n > 0 && right != NULL)
	{
		right = right->next;
		n--;
	}

	while (right != NULL)
	{
		left = left->next;
		right = right->next;
	}

	left->next = left->next->next;

	return dummyNode->next;
}

File: problems/LeetCode/linked_list/removeNthFromEnd.py
def removeNthFromEnd1(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
		if not head: return None

		length = 0
		cur = head
		# Get the length of the List
		while cur:
				length += 1
				cur = cur.next


		dummyNode = ListNode()
		dummyNode.next = head
		cur = dummyNode

		for i in range(0, length - n):
				cur = cur.next

		cur.next = cur.next.next


		return dummyNode.next


def removeNthFromEnd2(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
		left = right = dummyNode = ListNode(next = head)

		for _ in range(n):
				right = right.next
		
		while right.next:
				left = left.next
				right = right.next

		left.next = left.next.next
		return dummyNode.next